zjit/src/hir.rs

```diff
@@ -2065,9 +2072,77 @@ impl Function {
+        // missing location.
+        let mut assigned_in = vec![None; self.num_blocks()];
+        let rpo = self.rpo();
+        // Begin with every block having every variable defined, except for the entry block, which
+        // starts with nothing defined.
+        assigned_in[self.entry_block.0] = Some(InsnSet::with_capacity(self.insns.len()));
+        for &block in &rpo {
+            if block != self.entry_block {
+                let mut all_ones = InsnSet::with_capacity(self.insns.len());
+                all_ones.insert_all();
+                assigned_in[block.0] = Some(all_ones);
+            }
+        }
+        let mut worklist = VecDeque::with_capacity(self.num_blocks());
+        worklist.push_back(self.entry_block);
+        while let Some(block) = worklist.pop_front() {
+            let mut assigned = assigned_in[block.0].clone().unwrap();
+            for &param in &self.blocks[block.0].params {
+                assigned.insert(param);
+            }
+            for &insn_id in &self.blocks[block.0].insns {
+                let insn_id = self.union_find.borrow().find_const(insn_id);
+                match self.find(insn_id) {
+                    Insn::Jump(target) | Insn::IfTrue { target, .. } | Insn::IfFalse { target, .. } => {
+                        let Some(block_in) = assigned_in[target.target.0].as_mut() else {
+                            let fun_string = format!("{:?}", self);
+                            return Err(ValidationError::JumpTargetNotInRPO(fun_string, target.target));
+                        };
+                        // jump target's block_in was modified, we need to queue the block for processing.
+                        if block_in.intersect_with(&assigned) {
+                            worklist.push_back(target.target);
+                        }
+                    }
+                    _ => {
+                        assigned.insert(insn_id);
+                    }
+                }
+            }
+        }
+        // Check that each instruction's operands is assigned
+        for &block in &rpo {
+            let mut assigned = assigned_in[block.0].clone().unwrap();
+            for &param in &self.blocks[block.0].params {
+                assigned.insert(param);
+            }
+            for &insn_id in &self.blocks[block.0].insns {
+                let insn_id = self.union_find.borrow().find_const(insn_id);
+                let mut operands = VecDeque::new();
+                let insn = self.find(insn_id);
+                self.worklist_traverse_single_insn(insn.clone(), &mut operands);
+                for operand in operands {
+                    if !assigned.get(operand) {
+                        let fun_string = format!("{:?}", self);
+                        return Err(ValidationError::OperandNotDefined(fun_string, block, insn_id));
+                    }
+                }
+                assigned.insert(insn_id);
```

Comment:
Still only if `insn.has_output()`

Reply:
Good to add a test for using an instruction that has no output as an operand

Reply:
Ah, then it should be marked as having an output!

Reply:
That can/should be stacked as a separate PR

Reply:
Added a TODO and a temporary workaround.